### PR TITLE
Improved perimeter tracking

### PR DESCRIPTION
**Pull request by JamieStumpORNL**
Improved perimeter tracking. Can be improved further but good enough to deal with Mona Lisa case (todo later).
Improved compute by precomputing division.

### CODE REVIEW

File: src/DataStructs.h [hunk 1]
```diff
@@ -90,6 +90,7 @@ struct SimParams {
 	string tracking = "None";		// meltpool tracking mode
 	double dt = 1e-5;			// timestep
 	int out_freq = 1;		// output frequency
+	double radiusCheck = 1.0;		// The radius to be checking perimeter points from (different behavior is >1)
```
Comment (gknapp1) [inline src/DataStructs.h:93]: Typo in comment? `different` -> `default`?

File: src/Run.cpp [hunk 2]
```diff
@@ -432,7 +432,7 @@ void Run::Solidify_NoTracking(Grid& grid, const Simdat& sim) {
 
 		#pragma omp parallel for num_threads(sim.settings.thnum) schedule(dynamic,1+sim.domain.pnum/sim.settings.thnum/64)
 		for (int p = 0; p < sim.domain.pnum; p++) {
-			if (grid.Calc_T(t, nodes, sim, true, p) < sim.material.T_liq){
+			if (!(grid.Calc_T(t, nodes, sim, true, p) >= sim.material.T_liq)){
```
Comment (gknapp1) [inline src/Run.cpp:435]: Point of clarification, why switch from `if <` to `if not >=`?

File: src/Util.cpp [hunk 3]
```diff
@@ -42,9 +42,10 @@ void Util::AddToNodes(Nodes& nodes, const int_seg seg) {
 	nodes.xb.push_back(seg.xb);
 	nodes.yb.push_back(seg.yb);
 	nodes.zb.push_back(seg.zb);
-	nodes.phix.push_back(seg.phix);
-	nodes.phiy.push_back(seg.phiy);
-	nodes.phiz.push_back(seg.phiz);
+	// TODO::FEATURE
+	nodes.phix.push_back(1.0/seg.phix);
+	nodes.phiy.push_back(1.0/seg.phiy);
+	nodes.phiz.push_back(1.0/seg.phiz);
```
Comment (gknapp1) [inline src/Util.cpp:48]: Points of clarification, why the change to reciprocal of the variables and what is the TODO feature?

**Review by gknapp1 — COMMENTED** [review]
Left a few comments, the code looks good as far as I understand it. But I would like to understand/document why some of the changes were implemented.

File: src/DataStructs.h (hunk 1 above)
Comment (JamieStumpORNL) [reply src/DataStructs.h:93]: The whole "()" probably shouldn't be there at all. I plan on making the behavior different when >1 to make it more optimal but that is a TODO

File: src/Grid.cpp [hunk 4]
```diff
@@ -490,29 +482,26 @@ vector<vector<double>> Grid::Calc_Solidficiaton_Secondary(const double t, const
...
+		const double dpy = (ddpy * dy);
+		const double dpz = (ddpz * dz);
 
 		Gx_temp += dT_seg * dpx;	//x-gradient
 		Gy_temp += dT_seg * dpy;	//y-gradient
 		Gz_temp += dT_seg * dpz;	//z-gradient
 
 		Laplace += dT_seg * (dpx * dpx + dpy * dpy + dpz * dpz + ddpx + ddpy + ddpz); //laplacian
-		if (nodes.dtau[iter] == 0) { dT_t += phi; }  //Notice dtau is not used, that is because we are looking the instantaneous change at that time
+		dT_t += (nodes.dtau[iter]== 0) ? phi : 0;  //Notice dtau is not used, that is because we are looking the instantaneous change at that time
```
[the remark this answers is not in the document]
Comment (JamieStumpORNL) [reply src/Grid.cpp:504]: Should be something like "notice that the interval of time we are integrating over is of size/length"

File: src/Util.cpp (hunk 3 above)
Comment (JamieStumpORNL) [reply src/Util.cpp:48]: Precomputing the reciprocal means that the reciprocal is only computed once (all points share the same nodes). Division is ~10x more costly than multiplication so it reduces the total compute cycles needed. "TODO" can be removed. Note that I didn't' test it out but it should give the same results.

**Comment by streeve** [comment]
@JamieStumpORNL all makes sense - can you make those small edits? Then I think we can merge

**Review by streeve — APPROVED** [review]

**Review by gknapp1 — APPROVED** [review]
